Fix Quiz expanding between questions

itsyme opened the pull request:
**What is the purpose of this pull request?**

- [ ] Documentation update
- [x] Bug fix
- [ ] Feature addition or enhancement
- [ ] Code maintenance
- [ ] DevOps
- [ ] Improve developer experience
- [ ] Others, please explain:

<!--
  If this pull request is addressing an issue, link to the issue: #xxx

  If this pull request completely addresses an issue, use one of the closing keywords: "Fixes #xxx" or "Resolves #xxx"
  https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

  Otherwise, elaborate further on the rationale of this pull request as needed
-->

**Overview of changes:**
Resolves #2126. Forces intro/outro cards and questions in Quiz to wait for the previous card to leave before allowing the next card to enter.

**Anything you'd like to highlight/discuss:**
As a side effect of this PR, `gotoNextQuestion` in `Quiz.vue` had to be refactored into two functions,`gotoNextQuestion` and `showNextQuestion`, in order for the hiding of the current question and the showing of the next question to be called in two separate places. The transition duration was also reduced to keep the same effect as before.

**Testing instructions:**

<!--
  Any special testing instructions **not including** our automated tests.
-->

**Proposed commit message: (wrap lines at 72 characters)**
Fix expanding of Quiz during transition

Cards in Quiz expands during transition from one to the next due to
both cards being rendered in the Quiz.

This could be undesirable as it is an unexpected behaviour for users
and could lead to confusion when doing the Quiz.

Let's make the next card in the Quiz wait for the previous card to 
leave before entering.

This will prevent the expansion of the Quiz component as it is
guaranteed that the Quiz will only have one card rendered at any
one time.

<!--
  See this link for more info on how to write a good commit message:
  https://se-education.org/guides/conventions/git.html

  As best as possible, write a succinct commit title in 50 characters

|---------This is the width of 50 chars----------|
|-----------This is the width of 72 chars for your reference-----------|
-->

---

**Checklist:** :ballot_box_with_check:

<!-- Leave non-applicable items unchecked -->

- [ ] Updated the documentation for feature additions and enhancements
- [x] Added tests for bug fixes or features
- [x] Linked all related issues
- [ ] No unrelated changes <!-- It's tempting, but increases the reviewer's work, and really pollutes the commit history =( -->

<!--
  We'll try our best to get to your PR within a week.
  If we haven't gotten to it then, or if your pull request resolves an urgent item, feel free to give us a ping!
-->

yucheng11122017 commented on packages/vue-components/src/questions/Quiz.vue:
```diff
@@ -128,12 +128,13 @@ export default {
   },
   methods: {
     gotoNextQuestion() {
+      const DELAY = 525;
```
Might be better to define this outside the method since I see it also used inline 153 :)

itsyme replied:
Yep I agree! I was trying to do that under methods but only functions can be declared in methods. I also tried implementing a getDelay() function but eslint threw me an error so I decided on this instead! Maybe I can try putting it under a different hook. What do you think?

yucheng11122017 replied:
I thinking more along the lines of pulling out the constant DELAY with other constants like STATE_QUIZ_FRESH in line 96 so other methods can reference this as well. What do you think?

itsyme replied:
Ahh I see. I think thats a great idea! I will update that in the next commit!

yucheng11122017 commented on packages/vue-components/src/__tests__/Quiz.spec.js:
```diff
@@ -3,6 +3,12 @@ import Quiz from '../questions/Quiz.vue';
 import QOption from '../questions/QOption.vue';
 
+// Function to delay test until question is rendered
+export default function quizTimeout() {
```
I think the export default is unecessary since the function is only used in this file :)

itsyme replied:
Good catch! I was trying to export it from another utils.js file but decided not to in the end and forgot to delete it!

yucheng11122017 reviewed (commented):
Hi @itsyme, thanks for the work on this! 
I think the reason why the things in ```Quiz.spec.js.snap``` is being deleted is because it hasn't rendered yet after the triggers are triggered. I added ```await quizTimeOut()``` to lines 198 and 237 in Quiz.spec.js and it seems to work. 
Also I was wondering how did you get 525 for the value of delay?

itsyme commented:
Thanks @yucheng11122017! I got 525 by experimenting. I tried using 500 since the transitions were 0.5s long under `styles` -> `.question` in `Question.vue`, however it seemed to be too short so I increased it till it achieved the desired result!

itsyme commented:
> Hi @itsyme, thanks for the work on this! I think the reason why the things in `Quiz.spec.js.snap` is being deleted is because it hasn't rendered yet after the triggers are triggered. I added `await quizTimeOut()` to lines 198 and 237 in Quiz.spec.js and it seems to work. Also I was wondering how did you get 525 for the value of delay?

Thank you for the feedback! I added `await quizTimeOut()` to those two lines and it works perfectly now!

ang-zeyu reviewed (commented):
![flicker](https://user-images.githubusercontent.com/3306138/220922919-8051e941-23c9-499e-ad53-98e0c6160249.gif)

There are some flickering issues introduced by this change.

Presumably because the timeout is timed to be slightly after the transition actually ends.

Consider trying:
- something better guaranteed https://developer.mozilla.org/en-US/docs/Web/API/Element/transitionend_event (what Vue internally uses also actually)
- if that alone dosen't work.. (i have no time to browse the spec to check if the new card is introduced right before or after the transition ends...) consider if you really want to be attaching the new card only after the transition of the previous card has finished. (I am personally fine with this, though this changes the UX)

   I believe there are just 2 main cases to consider:
   1. New card's height > old card's height
   2. Old card's height > new card's height

   What can you do in these scenarios?

   The implementation details should be clear after.

itsyme commented:
Hi @ang-zeyu! Thank you for the feedback! I will try out your suggestions in the next commit!

itsyme commented:
I managed to overcome this issue by using `position: relative` for the animation when entering, this allows the incoming question to render in the same place as the previous question. However, since the intro and outro cards are not questions and were not animated, I had to add an animation to them in order to remove the expansion from intro -> first question and last question -> outro. I also realised that the progress bar is being blocked again during the animation from intro -> first question and outro -> first question (when retrying) and am currently working to fix that.

ang-zeyu commented:
Not sure if it has been updated but there is still some flicker currently:

![paneltransitionflicker](https://user-images.githubusercontent.com/3306138/221342246-6a44240f-d6bc-40a2-b10d-c662fe1a226a.gif)

itsyme commented:
Hi @ang-zeyu! Hmm that is odd. Since I made the change I have not observed this behaviour. May I know which browser you are using so that I may further investigate this issue? Thank you!

ang-zeyu commented:
> Hi @ang-zeyu! Hmm that is odd. Since I made the change I have not observed this behaviour. May I know which browser you are using so that I may further investigate this issue? Thank you!

Desktop chrome/firefox + mobile chrome on android, it occurs about 50% of the time. Unlike the gif it also occurs on mcq/radio questions (unrelated to a specific type of question).

lhw-1 commented:
I have tested this on Ubuntu 20.04, on Firefox 109.0.1 and Chromium 110.0.5481.100. Both seem to work fine without flickering.

jovyntls commented:
Chipping in here on a different environment, I'm getting flickering as well on MacOS with Chrome:

https://user-images.githubusercontent.com/61113575/221511067-76bd8233-6797-4e67-9bb4-8a70ddc16dff.mov

And MacOS with FireFox:

https://user-images.githubusercontent.com/61113575/221511904-3a63ecba-051e-4bfa-91f2-64b7d7b2f67f.mov

But it's on and off - can't reproduce it consistently.

jovyntls commented:
@itsyme Try setting throttling using Chrome dev tools (open a mobile view and adjust the throttling option)? I can reproduce the flickering on this branch much more easily when I set throttling to "low-end mobile" 
<img width="185" alt="image" src="https://user-images.githubusercontent.com/61113575/221520129-96b49383-208c-4483-8e43-c4bfe7ea69f6.png">


Uploading Screen Recording 2023-02-27 at 5.02.57 PM.mov…

itsyme commented:
Hi @lhw-1 @jovyntls! Thanks for testing it out! My guess is that the previous question's leave transition finishes slightly before the next question's enter transition finishes, leading to the div becoming height 0 for a brief moment. I will try to fix this else I will explore a different approach!

itsyme commented:
I managed to solve the issue by changing my approach. However as a side effect of this, `gotoNextQuestion` in `Quiz.vue` had to be refactored into two functions, `gotoNextQuestion` and `showNextQuestion`, in order for the hiding of the current question and the showing of the next question to be called in two separate places. Also, the animation duration was shortened in order to keep a similar effect to the previous animation, since the next question now waits for the previous question to fully leave before starting its animation.

ang-zeyu commented:
@itsyme Sorry for the late response.

This UX is completely fine to me, I don't see any flickering anymore as well.

> My guess is that the previous question's leave transition finishes slightly before the next question's enter transition finishes, leading to the div becoming height 0 for a brief moment. I will try to fix this else I will explore a different approach!

If you'd still like to experiment with the original UX outside of this PR, 

If I recall previously you were trying to:
- Introduce [absolute](https://github.com/MarkBind/markbind/pull/2184/commits/1a64885f47c244c399c7bc405ff57f3c0f1a078a#diff-3bcc5ce549c558f30435a786ffc4c2c80bc10e4415f2553083f5e354618bdbd1L188) positioning [before](https://vuejs.org/guide/built-ins/transition.html#css-based-transitions) the element is inserted.
- Remove the absolute positioning [after the transition](https://vuejs.org/guide/built-ins/transition.html#css-based-transitions) has finished


I think there's a few other options:

1. Let the quiz expand/shrink to the next card's height just as the "previous card's transition ends".  
   To be exact this can be either:
   - previous card's transition ends - it's still in the DOM so may not be the best option 
   - previous card leaves the DOM - worth a try if you can sync the removal of `position: absolute` on the next card to this.
   - the next card's transition ends - I believe were going for this previously, it is difficult since the previous card leaving the DOM is not synced to this.
2. A much easier option (I think it would be better UX wise as well) is to immediately let the quiz expand/shrink to the next card's height *just as* the user clicks "next". Because you can sync `position: absolute` on the previous card and the introduction of the next card right at the start.
3. Make all child cards `position:absolute` inside the quiz, manage the `height` at the `<quiz>` level through the transition.

ang-zeyu commented on packages/vue-components/src/__tests__/Quiz.spec.js:
```diff
@@ -51,6 +56,12 @@ const BLANKS_QUESTION = `
+// Function to delay test until question is rendered
+function timeoutRenderQuestion() {
+  // eslint-disable-next-line no-promise-executor-return
+  return new Promise(resolve => setTimeout(resolve, 350));
```
Is there no way to work around using a timeout?

Something like `waitForSelector(...).toGoMissing()` (imaginary).
Or creating your own simple transition stub.

There are 17 calls here so resolving this should help improve the speed of tests a fair bit. Unless the runner is parallel, I can't remember..

But don't spend too much time on it, 6s to me is still an ok degradation for now. (just create a follow up issue if we can't resolve this).

ang-zeyu replied:
@jovyntls could you help to merge in this MR if it looks good to you? 🙇

jovyntls reviewed (approved):
LGTM! Nice fix @itsyme, let me know if you'll be adding the tests now or opening a future issue for it - will merge if it's the latter

itsyme commented:
> LGTM! Nice fix @itsyme, let me know if you'll be adding the tests now or opening a future issue for it - will merge if it's the latter

I will open a new issue for this, thank you!
